### ISSUE TITLE
Can it run on CPU?

### ISSUE THREAD

**Issue by abdul-jabbar-ms** [jzhang38/TinyLlama#14]
Hello, can it run on cpu? with 4GB RAM. 
Can you please guide me regarding the minimum hardware requirements?

Thanks in advance.

**Comment by VatsaDev** [comment]
Check out #6  #11  for details. The first checkpoint **redlined** the colab GPU instance with 15GB
No way its running on 4GB CPU.

**Comment by abdul-jabbar-ms** [comment]
> Check out #6 #11 for details. The first checkpoint **redlined** the colab GPU instance with 15GB No way its running on 4GB CPU.

Thanks for the reply. I ran it on a CPU with 16 GB RAM, GitHub Code space. It worked. Though its answers were quite unrelated.

**Comment by ChaosCodes** [comment]
I think 4GB could work. Below is a screenshot of llama.cpp running on my Mac. It shows that `llm_load_tensors: mem required  =  606.59 MB (+   11.00 MB per state)`.
<img width="695" alt="image" src="https://github.com/jzhang38/TinyLlama/assets/23294676/8e07f5da-5dd2-4548-9a73-9e40f82afa63">
You can follow the instruction in [llama.cpp](https://github.com/ggerganov/llama.cpp) to try it on your CPU.

**Comment by VatsaDev** [comment]
@ChaosCodes is this for the 4bit quant version

also, How are you getting so much better performance on other devices? The T4 is redlined, yet a CPU works Just as well and you can run it at much less ram with llama.cpp?

is this an issue with the colab settings, or vLLM, or something else?

**Comment by Green-Sky** [comment]
the model is trained on **32bit floats**. most LLMs these days are trained on f16 or bf16, which cuts **memory usage in half**. you can convert it to 16bit floats yourself with llama.cpp
```
model size  =  4196.35 MB
quant size  =  2024.62 MB
```

```
llm_load_print_meta: format         = GGUF V2 (latest)
llm_load_print_meta: arch           = llama
llm_load_print_meta: vocab type     = SPM
llm_load_print_meta: n_vocab        = 32000
llm_load_print_meta: n_merges       = 0
llm_load_print_meta: n_ctx_train    = 2048
llm_load_print_meta: n_ctx          = 512
llm_load_print_meta: n_embd         = 2048
llm_load_print_meta: n_head         = 32
llm_load_print_meta: n_head_kv      = 4
llm_load_print_meta: n_layer        = 22
llm_load_print_meta: n_rot          = 64
llm_load_print_meta: n_gqa          = 8
llm_load_print_meta: f_norm_eps     = 1,0e-05
llm_load_print_meta: f_norm_rms_eps = 1,0e-05
llm_load_print_meta: n_ff           = 5632
llm_load_print_meta: freq_base      = 10000,0
llm_load_print_meta: freq_scale     = 1
llm_load_print_meta: model type     = ?B
llm_load_print_meta: model ftype    = mostly F16
llm_load_print_meta: model size     = 1,10 B
llm_load_print_meta: general.name   = models
llm_load_print_meta: BOS token = 1 '<s>'
llm_load_print_meta: EOS token = 2 '</s>'
llm_load_print_meta: UNK token = 0 '<unk>'
llm_load_print_meta: LF token  = 13 '<0x0A>'
llm_load_tensors: ggml ctx size =    0,06 MB
llm_load_tensors: mem required  = 2024,69 MB (+   11,00 MB per state)
........................................................................................
llama_new_context_with_model: kv self size  =   11,00 MB
llama_new_context_with_model: compute buffer total size =   67,97 MB
```

**Comment by Green-Sky** [comment]
also, because the model is using gqa, large contexts use almost no memory:
```
llama_new_context_with_model: kv self size  =   88,00 MB
llama_new_context_with_model: compute buffer total size =  269,47 MB
main: warning: model was trained on only 2048 context tokens (4096 specified)
```

**Comment by jzhang38** [comment]
@Green-Sky  Thanks for sharing your results! Just want to highlight that TinyLlama is trained with bf-16 mixed precision, the relevant code is [here](https://github.com/jzhang38/TinyLlama/blob/1e1d9ece6f17b8ea56afaca8f8c71651dccdd8fa/pretrain/tinyllama.py#L86).

**Comment by Green-Sky** [comment]
> is trained with bf-16 mixed precision

I stand corrected. I just thought it was, because you upload them in full 32bit.
... since bf16 is truncated f32, this means half of the model on hf is `0x0000` :smile:

**Comment by Green-Sky** [comment]
> Mixed precision training converts the weights to FP16 and calculates the gradients, before converting them back to FP32 **before multiplying by the learning** rate and updating the weights in the optimizer.

thanks, TIL
